### DISCUSSION TITLE
Git password authentication deprecated

### DISCUSSION THREAD

**Discussion by reikdas** [community/community#23219] [New to GitHub]
<p>Since using my GitHub password to push commits is deprecated - <a href="https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/" class="inline-onebox" rel="noopener nofollow ugc">Token authentication requirements for Git operations - The GitHub Blog</a> what is the optimal command line workflow now?</p>
<p>One of the recommended ways is to <a href="https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/#what-you-need-to-do-today" rel="noopener nofollow ugc">use a personal access token</a> but do I need to copy the generated token somewhere (since it is hidden in my GitHub profile after I generate it) and use it whenever I authenticate my password on a system?<br>
I added it to cache following the instructions here - <a href="https://docs.github.com/en/free-pro-team@latest/github/using-git/caching-your-github-credentials-in-git" class="inline-onebox" rel="noopener nofollow ugc">Caching your GitHub credentials in Git - GitHub Docs</a> but I am still asked for my password/access token when pushing to GitHub.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="reikdas" data-post="1" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 reikdas:</div>
<blockquote>
<p>One of the recommended ways is to <a href="https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/#what-you-need-to-do-today">use a personal access token</a> but do I need to copy the generated token somewhere (since it is hidden in my GitHub profile after I generate it) and use it whenever I authenticate my password on a system?</p>
</blockquote>
</aside>
<p>Yes, you’ll need the token where you previously typed your password. I recommend keeping the token(s) in a password manager that stores them in encrypted form.</p>
<aside class="quote no-group" data-username="reikdas" data-post="1" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 reikdas:</div>
<blockquote>
<p>I added it to cache following the instructions here - <a href="https://docs.github.com/en/free-pro-team@latest/github/using-git/caching-your-github-credentials-in-git">Caching your GitHub credentials in Git - GitHub Docs </a> but I am still asked for my password/access token when pushing to GitHub.</p>
</blockquote>
</aside>
<p>The memory cache described there keeps your password or token in RAM for a limited time, it’s not meant to store it permanently.</p>

**Comment by reikdas** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The memory cache described there keeps your password or token in RAM for a limited time, it’s not meant to store it permanently.</p>
</blockquote>
</aside>
<p>Is there a way to keep the token cached for a longer time period?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>The documentation you linked (<a href="https://docs.github.com/en/free-pro-team@latest/github/using-git/caching-your-github-credentials-in-git">Caching your GitHub credentials in Git</a>) describes how to change the timeout, but the RAM cache will <em>never</em> exist past something like a reboot. That’s a security feature.</p>
<p>For long term storage, use a password manager (I like <a href="https://keepassxc.org/">KeePassXC</a>, personally).</p>
<p>Technically Git also has <a href="https://git-scm.com/docs/git-credential-store">the “store” credential helper</a>, but note the warning there: :warning:</p>
<blockquote>
<p>Using this helper will store your passwords unencrypted on disk, protected only by filesystem permissions. If this is not an acceptable security tradeoff, try <a href="https://git-scm.com/docs/git-credential-cache">git-credential-cache[1]</a>, or find a helper that integrates with secure storage provided by your operating system.</p>
</blockquote>
<p>Which means in most cases it’s a bad idea, <em>maybe</em> with the exception of a machine that you use exclusively (NEVER anyone else) and that uses strong full-disk encryption.</p>

**Comment by reikdas** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>For long term storage, use a password manager (I like <a href="https://keepassxc.org/" rel="noopener nofollow ugc">KeePassXC</a>, personally).</p>
</blockquote>
</aside>
<p>I am not  a fan of password managers, so I am trying to avoid going down that route.</p>
<aside class="quote no-group quote-modified" data-username="airtower-luna" data-post="4" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Technically Git also has <a href="https://git-scm.com/docs/git-credential-store" rel="noopener nofollow ugc">the “store” credential helper</a>, but note the warning there: :warning:</p>
<blockquote>
<p>Using this helper will store your passwords unencrypted on disk, protected only by filesystem permissions. If this is not an acceptable security tradeoff, try <a href="https://git-scm.com/docs/git-credential-cache" rel="noopener nofollow ugc">git-credential-cache[1]</a>, or find a helper that integrates with secure storage provided by your operating system.</p>
</blockquote>
<p>Which means in most cases it’s a bad idea, <em>maybe</em> with the exception of a machine that you use exclusively (NEVER anyone else) and that uses strong full-disk encryption.</p>
</blockquote>
</aside>
<p>I am the exclusive user of my computer, but I am not very comfortable with having the token lay around on my system in cleartext (but this is what I am opting for till I figure out a permanent solution to this - Thanks!)</p>
<p>Is using ssh the only alternate left? I imagine that would allow me to enter my “GitHub” password once and not have to re-enter my password for that particular session?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="reikdas" data-post="5" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 reikdas:</div>
<blockquote>
<p>I am not very comfortable with having the token lay around on my system in cleartext</p>
</blockquote>
</aside>
<p>That’s exactly why I recommend a password manager. Not the kind that uploads everything, the good kind that stores your passwords in encrypted form. :wink:</p>
<aside class="quote no-group" data-username="reikdas" data-post="5" data-topic="153737">
<div class="title">
<div class="quote-controls"></div>
 reikdas:</div>
<blockquote>
<p>Is using ssh the only alternate left? I imagine that would allow me to enter my “GitHub” password once and not have to re-enter my password for that particular session?</p>
</blockquote>
</aside>
<p>SSH works only with public key authentication. You can store the key in encrypted form (similar to using a password manager) or unencrypted (similar to leaving the token on disk).</p>
